lib/src/clojure_lsp/feature/file_management.clj

```diff
@@ -46,58 +46,72 @@
         (async/>!! db/edits-chan (shared/client-changes changes db))))))
 
+(defn ^:private find-changed-elems-by
+  "Detect elements that changed number of occurrences."
+  [signature old-elems new-elems]
+  (comment
+    ;; increased
+    (merge-with - {:a 2} {:a 1}) ;; => {:a 1}
+    ;; decreased
+    (merge-with - {:a 1} {:a 2}) ;; => {:a -1}
+    ;; removed
+    (merge-with - {} {:a 1}) ;; => {:a 1} ;; not {:a -1}, as you'd expect with removing, but at least it's not 0
+    ;; added
+    (merge-with - {:a 1} {}) ;; => {:a 1}
+    ;; same
+    (merge-with - {:a 1} {:a 1}) ;; => {:a 0}
+    )
+  (let [signature-with-elem (fn [elem]
+                              (with-meta (signature elem) {:elem elem}))
+        old-counts (->> old-elems (map signature-with-elem) frequencies)
+        new-counts (->> new-elems (map signature-with-elem) frequencies)]
+    (->> (merge-with - new-counts old-counts)
+         (medley/remove-vals zero?)
+         keys
+         (map (comp :elem meta)))))
+
 (defn ^:private find-changed-var-definitions [old-local-analysis new-local-analysis]
   (let [old-var-defs (filter #(identical? :var-definitions (:bucket %)) old-local-analysis)
         new-var-defs (filter #(identical? :var-definitions (:bucket %)) new-local-analysis)
-        compare-fn (fn [other-var-defs {:keys [name fixed-arities]}]
-                     (if-let [var-def (first (filter #(= name (:name %)) other-var-defs))]
-                       (and (not (= fixed-arities (:fixed-arities var-def)))
-                            (not= 'clojure.core/declare (:defined-by var-def)))
-                       true))]
-    (->> (concat
-           (filter (partial compare-fn new-var-defs) old-var-defs)
-           (filter (partial compare-fn old-var-defs) new-var-defs))
-         (medley/distinct-by (juxt :name)))))
+        definition-signature (juxt :ns :name :fixed-arities :defined-by)]
+    (find-changed-elems-by definition-signature old-var-defs new-var-defs)))
 
 (defn ^:private find-changed-var-usages
   [old-local-analysis new-local-analysis]
   (let [old-var-usages (filter #(identical? :var-usages (:bucket %)) old-local-analysis)
         new-var-usages (filter #(identical? :var-usages (:bucket %)) new-local-analysis)
-        compare-fn (fn [other-var-usages current-var-usages var-usage]
-                     (= (count (filter #(= (:name var-usage) (:name %)) current-var-usages))
-                        (count (filter #(= (:name var-usage) (:name %)) other-var-usages))))]
-    (->> (concat
-           (remove (partial compare-fn new-var-usages old-var-usages) old-var-usages)
-           (remove (partial compare-fn old-var-usages new-var-usages) new-var-usages))
-         (medley/distinct-by (juxt :name)))))
+        usage-signature (juxt :to :name)]
+    (find-changed-elems-by usage-signature old-var-usages new-var-usages)))
 
 (defn ^:private notify-references [filename old-local-analysis new-local-analysis {:keys [db producer]}]
```

Comment:
we should probably add a test for this feature 😔 but I'm ok adding as a separated PR

Reply:
I know... I was kind of testing it in a REPL, but all the functions are private, meaning it's harder to test. Let's plan to do it in a separate PR.

Reply:
I'm totally ok testing private functions with `#'` [hack](https://github.com/clojure-lsp/clojure-lsp/blob/master/lib/test/clojure_lsp/shared_test.clj#L104)